local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,220 @@ fn test_slot_hash_expiry() {
     );
 }
 
+// This tests that when vote refresh should be sent, but the old vote is too old such that it's
+// outside slothash of the newest bank already so it will never land, we should send vote at the
+// tip of the current fork.
+// It creates validators A, B, and C, (C is only needed for stake distribution), then do this:
+// 1. kill B and C, let A run for a while to create its own fork A
+// 2. kill A, copy ledger of A to B so B has everything until the common ancestor, restart B and
+//    let it run long enough to create a new fork B
+// 3. bring back A and B only, A now has 37% of the stake so its fork is heavier, B has only 35%
+//    so it wants to switch to fork B, but can't do it because of A's stake under switch proof
+//    threshold of 38%. Also, B has voted on fork B, and last_voted < last_landed_vote (we do this
+//    by stopping B after its 4-block leader slot), and last vote is old enough so it's outside
+//    slothash
+// 4. we check that B will be able to vote again and the new vote is further down the B fork.
+#[test]
+#[serial]
+fn test_vote_refresh_outside_slothash() {
+    solana_logger::setup_with_default(RUST_LOG_FILTER);
+    solana_sdk::slot_hashes::set_entries_for_tests_only(8);
+
+    let slots_per_epoch = 2048;
+    let node_stakes = vec![
+        37 * DEFAULT_NODE_STAKE,
+        35 * DEFAULT_NODE_STAKE,
+        28 * DEFAULT_NODE_STAKE,
+    ];
+    let validator_keys = vec![
+        "28bN3xyvrP4E8LwEgtLjhnkb7cY4amQb6DrYAbAYjgRV4GAGgkVM2K7wnxnAS7WDneuavza7x21MiafLu1HkwQt4",
+        "2saHBBoTkLMmttmPQP8KfBkcCw45S5cwtV3wTdGCscRC8uxdgvHxpHiWXKx4LvJjNJtnNcbSv5NdheokFFqnNDt8",
+        "4mx9yoFBeYasDKBGDWCTWGJdWuJCKbgqmuP8bN9umybCh5Jzngw7KQxe99Rf5uzfyzgba1i65rJW4Wqk7Ab5S8ye",
+    ]
+    .iter()
+    .map(|s| (Arc::new(Keypair::from_base58_string(s)), true))
+    .collect::<Vec<_>>();
+    let node_vote_keys = vec![
+        "3NDQ3ud86RTVg8hTy2dDWnS4P8NfjhZ2gDgQAJbr3heaKaUVS1FW3sTLKA1GmDrY9aySzsa4QxpDkbLv47yHxzr3",
+        "46ZHpHE6PEvXYPu3hf9iQqjBk2ZNDaJ9ejqKWHEjxaQjpAGasKaWKbKHbP3646oZhfgDRzx95DH9PCBKKsoCVngk",
+        "3zsEPEDsjfEay7te9XqNjRTCE7vwuT6u4DHzBJC19yp7GS8BuNRMRjnpVrKCBzb3d44kxc4KPGSHkCmk6tEfswCg",
+    ]
+    .iter()
+    .map(|s| Arc::new(Keypair::from_base58_string(s)))
+    .collect::<Vec<_>>();
+    let vs = validator_keys
+        .iter()
+        .map(|(kp, _)| kp.pubkey())
+        .collect::<Vec<_>>();
+    let (a_pubkey, b_pubkey, c_pubkey) = (vs[0], vs[1], vs[2]);
+
+    // We want B to not vote (we are trying to simulate its votes not landing until it gets to the
+    // minority fork)
+    let mut validator_configs =
+        make_identical_validator_configs(&ValidatorConfig::default_for_test(), node_stakes.len());
+    validator_configs[1].voting_disabled = true;
+
```

Comment:
it would be nice to setup a custom leader schedule so we don't have to wait to skip B's leader slots while it is offline or vice versa when A is offline.

Reply:
Done.